### PR TITLE
Fix comment in FSDP

### PR DESCRIPTION
**Pull request by rohan-varma**
Addresses @awgu comment on wording this comment better.

### CODE REVIEW

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 1]
```diff
@@ -3134,10 +3134,13 @@ def _finalize_params(fsdp_module: FullyShardedDataParallel) -> None:
...
-                            # did not have a gradient computed. It was either unused
-                            # in forward, or unused in loss computation so it did
-                            # not get gradient
+                            # no FSDP managed param's post_backward_hook
+                            # was called, meaning that no FSDP managed param
+                            # had a gradient computed. This means that module
+                            # was either unused in forward or module output
+                            # was unused in loss computation, so expected
+                            # state is either backward_pre or idle depending
+                            # on if pre-backward hook was run.
```
Comment (awgu) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:3143]: I prefer us to be explicit on why a pre-backward hook would run or not. Reading this comment, I am not entirely sure how the preconditions of {module unused in forward pass, module output unused in loss computation} map to {`IDLE`, `BACKWARD_PRE`}.

See the overall PR review comment for more context.

**Review by awgu — COMMENTED** [review]
This is not directly related to the PR or your recent fix, but I think we need to revisit this entire block:
https://github.com/pytorch/pytorch/blob/8786ca0140bcae6219095b853bf090eff13f0fed/torch/distributed/fsdp/fully_sharded_data_parallel.py#L3121-L3155

1. If a module is used in the forward pass where the forward output tensors require gradients, then a pre-backward hook is registered to the output tensors for that module's `FlatParameter`.
    1. If any of the output tensors are used in the loss computation, then `loss.backward()` computes a gradient with respect to the output tensor(s) and runs the pre-backward hook. The module's training state is at least `PRE_BACKWARD`.
        1. If the a module's `FlatParameter` was used to compute an output tensor receiving a gradient, then the `FlatParameter` receives a gradient, and the post-backward hook runs. The training state is `POST_BACKWARD`.
        1. If none of the module's `FlatParameter`s were used to compute any of the output tensors receiving a gradient, then none of the `FlatParameter`s receive a gradient, and the post-backward hook does not run. The training state is `PRE_BACKWARD`.
    2. If none of the output tensors are used in the loss computation, then `loss.backward()` will not compute a gradient with respect to the output tensors, and the pre-backward hook does not run. The module's training state is `IDLE`.
2. If a module is not used in the forward pass or the output tensors do not require gradients, then no pre-backward hook is registered for that module's `FlatParameter`, and the training state of that module is `IDLE`.
    1. If the module is not used in the forward pass, then no post-backward hook is registered.
    2. If the module is used in the forward pass but none of the output tensors require gradients, then the post-backward hooks never run.
---
Some new thoughts today:

- Before https://github.com/pytorch/pytorch/pull/80245/, we asserted the training state based on "external information", namely based on parameters' `.requires_grad` (all nested parameters and only directly managed parameters).
- After https://github.com/pytorch/pytorch/pull/80245/, we have a branch that asserts the training state based on "internal information": We directly set a flag in the post-backward hook to indicate that it ran and later assert that if that flag is set, then the training state is `BACKWARD_POST`.
    - IIUC, this is actually not a meaningful assert. This is checking if we set `self.training_state` correctly. In other words, if we set `_post_backward_hook = True` *and* `self.training_state = BACKWARD_POST` in `_post_backward_hook()`, then this check is not doing much.

If anything, what we should do is track the `requires_grad` status of the output tensors in the module's `forward()`, and call asserts based on that and the parameter `requires_grad` status. It makes sense to only use "external information" to check that our understanding of the autograd engine matches reality.

---
I am good with landing this PR in the interim, and I can follow up with sharpening these assertions. Before I approve though, I would love to hear everyone's thoughts.

cc: @zhaojuanmao

**Comment by rohan-varma** [comment]
@awgu  I agree we should try to use external information, i.e. from the autograd engine only. However, the original code verified that the training state was in backward post which was set in post backward hook, AFAIK there is no way to use only external information to ensure post backward hook was called? If we really want, we could somehow intercept the loss tensor (don't think this is possible today) and do a graph traversal to truly figure out which parameters will get gradient, but other than that I'm not sure there's a great way to figure out if post backward hook was fired.

**Comment by awgu** [comment]
> @awgu I agree we should try to use external information, i.e. from the autograd engine only. However, the original code verified that the training state was in backward post which was set in post backward hook, AFAIK there is no way to use only external information to ensure post backward hook was called? If we really want, we could somehow intercept the loss tensor (don't think this is possible today) and do a graph traversal to truly figure out which parameters will get gradient, but other than that I'm not sure there's a great way to figure out if post backward hook was fired.

If we assert that the training state is `BACKWARD_POST`, then we know that the post-backward hook was called. It is possible that if we additionally track whether pre-backward hooks were registered (iff output tensors required gradients) on top of the the existing parameter `requires_grad` checks, then we can cover all cases where we expect the training state to be `IDLE` or `BACKWARD_PRE`. Then, the complement (e.g. an `else` branch) leaves all cases where we expect `BACKWARD_POST`. Let me revisit this later after I have put out the non-recursive wrapping PR.
